### PR TITLE
#116 Relative stability

### PR DESCRIPTION
**Pull request by aaarrti**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,6 +112,9 @@ measures to what extent explanations are stable when subject to slight perturbat
     <li><b>Max-Sensitivity </b><a href="https://arxiv.org/pdf/1901.09392.pdf">(Yeh et al., 2019)</a>: measures the maximum sensitivity of an explanation using a Monte Carlo sampling-based approximation
     <li><b>Avg-Sensitivity </b><a href="https://arxiv.org/pdf/1901.09392.pdf">(Yeh et al., 2019)</a>: measures the average sensitivity of an explanation using a Monte Carlo sampling-based approximation
     <li><b>Continuity </b><a href="https://arxiv.org/pdf/1706.07979.pdf">(Montavon et al., 2018)</a>: captures the strongest variation in explanation of an input and its perturbed version
+    <li><b>Relative Input Stability (RIS)</b><a href="https://arxiv.org/pdf/2203.06877.pdf"> (Chirag Agarwal, et. al., 2022)</a>: measures the relative distance between explanations e_x and e_x' with respect to the distance between the two inputs x and x'
```
Comment (annahedstroem) [inline README.md:115]: could be good to mention something around representations? that x can be any layer e.g., an input- or hidden layer of the network

File: .github/workflows/python-package.yml [hunk 2]
```diff
@@ -16,7 +16,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.7", "3.8", "3.9"]
+        python-version: ["3.8", "3.9"]
```
Comment (annahedstroem) [inline .github/workflows/python-package.yml:19]: Why this?

File: quantus/helpers/model_interface.py [hunk 3]
```diff
@@ -43,3 +43,16 @@ def get_random_layer_generator(self):
         Set order to independent for independent randomization.
         """
         raise NotImplementedError
+
```
Comment (annahedstroem) [inline quantus/helpers/model_interface.py:46]: Maybe refactor names to `get_hidden_layers_representations` or `get_hidden_layers_activations`?

File: quantus/helpers/models.py [hunk 4]
```diff
@@ -157,3 +157,31 @@ def __init__(self, n_channels, seq_len, n_classes):
...
             )
+
+
+    def cnn_2d_tf(
+            img_height: int,
+            img_width: int,
+            num_classes: int,
+            num_channels=3
+    ) -> tf.keras.Model:
+        # NN adapted from https://www.tensorflow.org/tutorials/images/classification
```
Comment (annahedstroem) [inline quantus/helpers/models.py:168]: Proper docstring please!

File: quantus/helpers/models.py [hunk 5]
```diff
@@ -157,3 +157,31 @@ def __init__(self, n_channels, seq_len, n_classes):
                 loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                 metrics=[tf.keras.metrics.SparseCategoricalAccuracy()],
             )
+
+
+    def cnn_2d_tf(
```
Comment (annahedstroem) [inline quantus/helpers/models.py:162]: Rename CNN_2D_TF

File: quantus/helpers/perturb_func.py [hunk 6]
```diff
@@ -389,3 +389,9 @@ def noisy_linear_imputation(
 def no_perturbation(arr: np.array, **kwargs) -> np.array:
     """Apply no perturbation to input."""
     return arr
+
+
+def random_noise(arr: np.ndarray, upper_bound=1e-6, **kwargs) -> np.ndarray:
```
Comment (annahedstroem) [inline quantus/helpers/perturb_func.py:394]: One sentence docstring, pls.

File: quantus/helpers/pytorch_model.py [hunk 7]
```diff
@@ -87,3 +87,31 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
             torch.manual_seed(seed=seed + 1)
             module[1].reset_parameters()
             yield module[0], random_layer_model
+
+    def get_hidden_layers_outputs(self, x: np.ndarray,
+                                  layer_names: Optional[List[str]] = None,
+                                  layer_indices: Optional[List[int]] = None,
+                                  ) -> np.ndarray:
+        # skip last layer, and module defined by subclassing API
```
Comment (annahedstroem) [inline quantus/helpers/pytorch_model.py:95]: Missing docstring, please go over all functions and add where necessary.

File: quantus/helpers/pytorch_model.py [hunk 8]
```diff
@@ -87,3 +87,31 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
...
+        ))
+        if layer_names:
+            hidden_layers = list(filter(lambda l: l[0] in layer_names, hidden_layers))
+        if layer_indices:
+            hidden_layers = [hidden_layers[i] for i in layer_indices]
+
+        batch_size = x.shape[0]
+        hidden_outputs = []
+
+        def hook(module, module_in, module_out):
```
Comment (annahedstroem) [inline quantus/helpers/pytorch_model.py:109]: Please add comments on the below code!

File: quantus/helpers/tf_model.py [hunk 9]
```diff
@@ -9,6 +9,7 @@
 
 from ..helpers.model_interface import ModelInterface
 from ..helpers import utils
+import tensorflow as tf
```
Comment (annahedstroem) [inline quantus/helpers/tf_model.py:12]: Move above the local imports

File: quantus/helpers/tf_model.py [hunk 10]
```diff
@@ -93,3 +94,29 @@ def get_random_layer_generator(self, order: str = "top_down", seed: int = 42):
             np.random.seed(seed=seed + 1)
             layer.set_weights([np.random.permutation(w) for w in weights])
             yield layer.name, random_layer_model
+
+    def get_hidden_layers_outputs(self,
+                                  x: np.ndarray,
+                                  layer_names: Optional[List[str]] = None,
+                                  layer_indices: Optional[List[int]] = None,
+                                  ) -> np.ndarray:
+        batch_size = x.shape[0]
```
Comment (annahedstroem) [inline quantus/helpers/tf_model.py:103]: Docstring

File: quantus/metrics/robustness_metrics/__init__.py [hunk 11]
```diff
@@ -0,0 +1,2 @@
+from .robustness_metrics import *
+from .relative_stability import *
```
Comment (annahedstroem) [inline quantus/metrics/robustness_metrics/__init__.py:2]: Please add relative stability code to robustness_metrics.py (we might separate metrics out at a later point, but then that is a separate PR)

File: quantus/metrics/robustness_metrics/relative_stability.py [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+import jax
+from abc import abstractmethod, ABC
+import functools
+from tqdm import tqdm
+
+from ..base import Metric
+from ...helpers import utils, perturb_func
+
+"""
+For example usage, refer to tutorials/tutorial_getting_started_with_tensorflow.ipynb
```
Comment (annahedstroem) [inline quantus/metrics/robustness_metrics/relative_stability.py:13]: Please add to `tutorial_basic_all_metrics.ipynb`

Comment (annahedstroem) [reply quantus/metrics/robustness_metrics/relative_stability.py:13]: (and remove this line)

File: requirements_test.txt [hunk 13]
```diff
@@ -2,20 +2,21 @@ captum==0.4.1
...
 protobuf~=3.19.0
 pytest==6.2.5
 pytest-cov==3.0.0
 pytest-lazy-fixture==0.6.3
 scikit-image==0.19.1 ; python_version > '3.6'
-scikit-image==0.17.2 ; python_version == '3.6'
 scikit-learn==0.24.2
 scipy==1.5.4
-tensorflow==2.6.2
+tensorflow==2.9.1
```
Comment (annahedstroem) [inline requirements_test.txt:14]: Why is only `requirements_test.txt` and not `requirements.txt` updated?

Comment (annahedstroem) [reply requirements_test.txt:14]: Don't we e.g., need jax for both (also not sure about whether we want to include jax but let's discuss for performance gains )

File: quantus/metrics/robustness_metrics/robustness_metrics.py [hunk 14]
```diff
@@ -1,22 +1,21 @@
...
-from ..helpers import perturb_func
-from ..helpers import similar_func
-from ..helpers import utils
-from ..helpers import warn_func
-from ..helpers.asserts import attributes_check
-from ..helpers.model_interface import ModelInterface
-from ..helpers.norm_func import fro_norm
-from ..helpers.normalise_func import normalise_by_negative
-from ..helpers.discretise_func import top_n_sign
+from ..base import Metric
```
Comment (annahedstroem) [inline quantus/metrics/robustness_metrics/robustness_metrics.py:8]: Why is this changed? It is working as is

File: quantus/metrics/robustness_metrics/relative_stability.py [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+relative_stability_objective_batched = jax.vmap(
+    relative_stability_objective_single_point, in_axes=(0, 0, 0, 0, None, None)
+)
+
+relative_stability_objective_vectorized_over_perturbation_axis = jax.vmap(
+    relative_stability_objective_batched, in_axes=(None, 0, None, 0, None, None)
+)
+
+
+class RelativeStability(Metric, ABC):
```
Comment (annahedstroem) [inline quantus/metrics/robustness_metrics/relative_stability.py:44]: Inherit from Metric (only)

File: quantus/metrics/robustness_metrics/relative_stability.py [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+
+from ..base import Metric
+from ...helpers import utils, perturb_func
+
+"""
+For example usage, refer to tutorials/tutorial_getting_started_with_tensorflow.ipynb
+"""
+
+
+@functools.partial(jax.jit, static_argnums=(4, 5))
```
Comment (annahedstroem) [inline quantus/metrics/robustness_metrics/relative_stability.py:17]: Explain what is happening lines 17-41

File: quantus/metrics/robustness_metrics/relative_stability.py [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+
+        if "explain_func" in kwargs:
+            explain_func: Callable = kwargs.pop("explain_func")
+
+            a_batch = explain_func(
+                model=self.model.get_model(), inputs=x_batch, targets=y_batch, **kwargs
+            )
+
+            it = xs_batch
+            if self.display_progressbar:
```
Comment (annahedstroem) [inline quantus/metrics/robustness_metrics/relative_stability.py:198]: Where is this initalised?

File: quantus/metrics/robustness_metrics/relative_stability.py [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+        else:
+            a_batch = kwargs.pop("a_batch")
+            as_batch = kwargs.pop("as_batch")
+
+            if len(as_batch.shape) <= len(a_batch.shape):
+                raise ValueError(
+                    "Batch of perturbed explanations must have 1 more axis"
+                )
+
+        if self.normalise:
```
Comment (annahedstroem) [inline quantus/metrics/robustness_metrics/relative_stability.py:217]: Missing at init?

File: setup.py [hunk 19]
```diff
@@ -53,15 +53,15 @@ def version(s: str):
 # Define setup.
 setup(
     name="quantus",
-    version="0.1.5",
+    version="0.1.6",
```
Comment (annahedstroem) [inline setup.py:56]: I'll do a new release including some other changes so please remove

File: setup.py [hunk 20]
```diff
@@ -81,6 +83,6 @@ def version(s: str):
     license="CC BY-NC-SA 3.0",
     packages=find_packages(),
     zip_safe=False,
-    python_requires=">=3.7",
+    python_requires=">=3.8",
```
Comment (annahedstroem) [inline setup.py:86]: To be discussed!

File: tests/fixtures.py [hunk 21]
```diff
@@ -1,17 +1,9 @@
...
-import torchvision
-from torchvision import transforms
 import numpy as np
-from ..quantus.helpers.models import LeNet, LeNetTF, ConvNet1D, ConvNet1DTF
-from tensorflow.keras.models import load_model
-from tensorflow.keras.datasets import cifar10
-import tensorflow as tf
-from ..quantus.helpers.pytorch_model import PyTorchModel
-from ..quantus.helpers.tf_model import TensorFlowModel
+from ..quantus.helpers.models import LeNet, LeNetTF, ConvNet1D, ConvNet1DTF, cnn_2d_tf
```
Comment (annahedstroem) [inline tests/fixtures.py:5]: Naming convention!

File: tests/helpers/test_perturb_func.py [hunk 22]
```diff
@@ -508,3 +508,11 @@ def test_no_perturbation(
 ):
     out = no_perturbation(arr=data, **params)
     assert (out == data).all() == expected, "Test failed."
+
+
+@pytest.mark.perturb_func
+@pytest.mark.parametrize("data", [(lazy_fixture("load_mnist_images_tf"))])
```
Comment (annahedstroem) [inline tests/helpers/test_perturb_func.py:514]: Also try with PyTorch data please

File: tests/helpers/test_pytorch_model.py [hunk 23]
```diff
@@ -157,3 +150,28 @@ def test_get_random_layer_generator(load_mnist_model):
...
+@pytest.mark.parametrize(
+    "params", [
+        {},
+        {'layer_names': ['conv_2']},
+        {'layer_indices': [0, 1]}
+    ],
+    ids=[
+        'all layers',
+        '2nd conv',
+        '1st 2 layers'
```
Comment (annahedstroem) [inline tests/helpers/test_pytorch_model.py:165]: Please go back to get_hidden_layers and explain the options (str) clearly in the docstring!

**Review by annahedstroem — CHANGES_REQUESTED** [review]
Great work with the implementation! Also, great suggestion using jax. 

Please see comments.

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 18 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/relative_stability.py:217]: There's a call to __init__ of superclass (`quantus.Metric`), 
```
    def __init__(self, *args, **kwargs):
        """
        ...
        """
        super().__init__(*args, **kwargs)
        ...

 ```
 And `quantus.Metric` initialises this property
https://github.com/understandable-machine-intelligence-lab/Quantus/blob/main/quantus/metrics/base.py#L43

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 17 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/relative_stability.py:198]: There's a call to init of superclass (quantus.Metric),

    def __init__(self, *args, **kwargs):
        """
        ...
        """
        super().__init__(*args, **kwargs)
        ...

And quantus.Metric initialises this property
https://github.com/understandable-machine-intelligence-lab/Quantus/blob/main/quantus/metrics/base.py#L43

File: tests/helpers/test_pytorch_model.py (hunk 23 above)
Comment (aaarrti) [reply tests/helpers/test_pytorch_model.py:165]: done https://github.com/aaarrti/Quantus/blob/relative-stability/quantus/helpers/model_interface.py#L53-L68

File: quantus/helpers/model_interface.py (hunk 3 above)
Comment (aaarrti) [reply quantus/helpers/model_interface.py:46]: done

File: tests/helpers/test_pytorch_model.py (hunk 23 above)
Comment (aaarrti) [reply tests/helpers/test_pytorch_model.py:165]: done here https://github.com/aaarrti/Quantus/blob/relative-stability/quantus/helpers/model_interface.py#L53-L68

File: quantus/helpers/models.py (hunk 5 above)
Comment (aaarrti) [reply quantus/helpers/models.py:162]: done

File: quantus/helpers/models.py (hunk 4 above)
Comment (aaarrti) [reply quantus/helpers/models.py]: done https://github.com/aaarrti/Quantus/blob/relative-stability/quantus/helpers/models.py#L168-L170

File: quantus/metrics/robustness_metrics/__init__.py (hunk 11 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/__init__.py]: done

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 15 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/relative_stability.py]: without explicitly specifying ABC in inheritance list, user will be able to instantiate `quantus.RelativeStability`
class, which is not intended.

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 18 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/relative_stability.py]: there's also a test case for this https://github.com/aaarrti/Quantus/blob/relative-stability/tests/metrics/test_relative_stability.py#L379

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 17 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/relative_stability.py]: there's also a test case for this https://github.com/aaarrti/Quantus/blob/relative-stability/tests/metrics/test_relative_stability.py#L379

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 15 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/relative_stability.py]: there's a test case for this https://github.com/aaarrti/Quantus/blob/relative-stability/tests/metrics/test_relative_stability.py#L43

File: tests/fixtures.py (hunk 21 above)
Comment (aaarrti) [reply tests/fixtures.py]: changed

File: setup.py (hunk 19 above)
Comment (aaarrti) [reply setup.py]: reverted

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 16 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/relative_stability.py]: I have added as much explanation in comments, doc strings as I could think of https://github.com/aaarrti/Quantus/blob/relative-stability/quantus/helpers/relative_stability_objective.py

File: quantus/metrics/robustness_metrics/robustness_metrics.py (hunk 14 above)
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/robustness_metrics.py]: reverted

File: quantus/helpers/tf_model.py (hunk 10 above)
Comment (aaarrti) [reply quantus/helpers/tf_model.py:103]: there's one in base class https://github.com/aaarrti/Quantus/blob/relative-stability/quantus/helpers/model_interface.py#L53-L68.

File: quantus/helpers/pytorch_model.py (hunk 8 above)
Comment (aaarrti) [reply quantus/helpers/pytorch_model.py:114]: done

File: quantus/helpers/pytorch_model.py (hunk 7 above)
Comment (aaarrti) [reply quantus/helpers/pytorch_model.py]: there's one in base class https://github.com/aaarrti/Quantus/blob/relative-stability/quantus/helpers/model_interface.py#L53-L68.

File: quantus/helpers/perturb_func.py (hunk 6 above)
Comment (aaarrti) [reply quantus/helpers/perturb_func.py:394]: done https://github.com/aaarrti/Quantus/blob/relative-stability/quantus/helpers/perturb_func.py#L395-L402

File: requirements_test.txt (hunk 13 above)
Comment (aaarrti) [reply requirements_test.txt:14]: I did not see `requirements.txt` being used by anything. I will also update it

File: setup.py (hunk 20 above)
Comment (aaarrti) [reply setup.py:86]: reverted

File: requirements_test.txt (hunk 13 above)
Comment (aaarrti) [reply requirements_test.txt]: updated

File: quantus/metrics/robustness_metrics/relative_stability.py [hunk 24]
```diff
@@ -0,0 +1,312 @@
+from typing import Union, Dict, Callable, Tuple
```
[the remark this answers is not in the document]
Comment (aaarrti) [reply quantus/metrics/robustness_metrics/relative_stability.py]: done

File: tests/helpers/test_perturb_func.py (hunk 22 above)
Comment (aaarrti) [reply tests/helpers/test_perturb_func.py:514]: done

File: .github/workflows/python-package.yml (hunk 2 above)
Comment (annahedstroem) [reply .github/workflows/python-package.yml]: @aaarrti

File: requirements_test.txt (hunk 13 above)
Comment (annahedstroem) [reply requirements_test.txt]: ok thanks

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 15 above)
Comment (annahedstroem) [reply quantus/metrics/robustness_metrics/relative_stability.py]: why not instantiate `quantus.RelativeStability`?

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 12 above)
Comment (annahedstroem) [reply quantus/metrics/robustness_metrics/relative_stability.py]: @aaarrti

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 16 above)
Comment (annahedstroem) [reply quantus/metrics/robustness_metrics/relative_stability.py]: great looks good thanks.

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 17 above)
Comment (annahedstroem) [reply quantus/metrics/robustness_metrics/relative_stability.py]: FYI we are working on another PR that later will update this logic

but for now we want to follow the same syntax as the other metrics, as it needs to be able to be updated explicitly when a user passes the kwargs

I am aware that it is not optimal, and that your suggestion to not explicitly include it works, but for now, pls follow the syntax

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (annahedstroem) [reply quantus/metrics/robustness_metrics/relative_stability.py]: thanks!

**Comment by annahedstroem** [comment]
OK, we need to discuss the performance gains / pros and cons with jax, separately

File: quantus/metrics/robustness_metrics/relative_stability.py (hunk 18 above)
Comment (annahedstroem) [reply quantus/metrics/robustness_metrics/relative_stability.py]: see answer above!

File: tests/fixtures.py (hunk 21 above)
Comment (annahedstroem) [reply tests/fixtures.py]: should be CNN2DTF

File: tests/helpers/test_pytorch_model.py (hunk 23 above)
Comment (annahedstroem) [reply tests/helpers/test_pytorch_model.py:165]: 100%!

**Comment by aaarrti** [comment]
> OK, we need to discuss the performance gains / pros and cons with jax, separately

btw, I have made a small experiment comparing the performance https://github.com/aaarrti/Quantus/blob/jax-vs-np/notebooks/jax_vs_numpy_performance_comparison.ipynb

**Comment by aaarrti** [comment]
> OK, we need to discuss the performance gains / pros and cons with jax, separately

as discussed on 08.09.22, no `jax` for now. I've changed the implementation to use `numpy` only

**Comment by annahedstroem** [comment]
@aaarrti please resolve these conflicts and then I'll review the PR as requested. 

Please also add a description of the PR (just open up a new PR request and you'll get a template automatically populated)

**Comment by aaarrti** [comment]
It is easier to open a new one
